### PR TITLE
Geometry refactor

### PR DESCRIPTION
**Pull request by cboesch-gpsw**

### CODE REVIEW

File: libnodegl/geometry.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+#ifndef GEOMETRY_H
+#define GEOMETRY_H
+
+#include "buffer.h"
+
+struct gpu_ctx;
+
+int ngli_geometry_gen_vec3(struct buffer **bufferp, struct buffer_layout *layout,
```
Comment (mbouron-gpsw) [inline libnodegl/geometry.h:29]: `const struct buffer_layout *layout` ?

File: libnodegl/geometry.c [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "format.h"
+#include "geometry.h"
+#include "nodegl.h"
+#include "type.h"
+
+static int geometry_gen_buffer(struct buffer **bufferp, struct buffer_layout *layout,
```
Comment (mbouron-gpsw) [inline libnodegl/geometry.c:27]: `const struct buffer_layout *layout` ?

Comment (cboesch-gpsw) [reply libnodegl/geometry.c]: Fixed, thanks

File: libnodegl/node_geometry.c [hunk 3]
```diff
@@ -108,36 +108,82 @@ static const struct node_param geometry_params[] = {
     }                                                      \
 } while (0)                                                \
 
+static int configure_buffer(struct ngl_node *buffer_node, struct buffer **bufferp, struct buffer_layout *layout, int usage)
```
Comment (mbouron-gpsw) [inline libnodegl/node_geometry.c:84]: Please put `usage` as the second argument so all the outputs are at the end of the argument list.

Comment (cboesch-gpsw) [reply libnodegl/node_geometry.c]: Done.

**Comment by mbouron-gpsw** [comment]
Looks good to me, thanks !

**Review by mbouron-gpsw — APPROVED** [review]

File: libnodegl/buffer.h [hunk 4]
```diff
@@ -46,4 +46,13 @@ int ngli_buffer_init(struct buffer *s, int size, int usage);
...
 void ngli_buffer_freep(struct buffer **sp);
 
+/* helper structure to specify the content (or a sub-part) of a buffer */
+struct buffer_layout {
+    int type;       // any of NGLI_TYPE_*
+    int format;     // any of NGLI_FORMAT_*
+    int stride;     // stride of 1 element, in bytes
+    int comp;       // number of components per element
+    int count;      // number of elements
+};
```
[the remark this answers is not in the document]
Comment (cboesch-gpsw) [reply libnodegl/buffer.h]: Done.

**Pull request merged**